lib/schedule/swing.ex

```diff
@@ -116,4 +116,28 @@ defmodule Schedule.Swing do
     do: nil
 
   defp trip_or_trip_id_to_trip(trip_id, trips_by_id), do: Map.fetch!(trips_by_id, trip_id)
+
+  @spec first_trip_from_piece(Schedule.Piece.t(), Trip.by_id()) ::
+          Trip.id() | Schedule.Trip.t() | Schedule.AsDirected.t()
+  defp first_trip_from_piece(piece, trips_by_id) do
+    first_full_trip =
+      piece.trips
+      |> Stream.map(&trip_or_trip_id_to_trip(&1, trips_by_id))
+      |> Stream.filter(fn trip -> !is_nil(trip) end)
+      |> Enum.take(1)
+      |> List.first()
+
+    case {first_full_trip, piece.start_mid_route?} do
+      {nil, %{trip: mid_route_trip}} ->
+        mid_route_trip |> trip_or_trip_id_to_trip(trips_by_id)
+
+      {_, %{trip: mid_route_trip}} ->
+        with %{} = first_trip <- mid_route_trip |> trip_or_trip_id_to_trip(trips_by_id) do
+          %{first_trip | :run_id => first_full_trip.run_id}
```

Comment:
You can use the slightly more compact map syntax when the key is an atom.

```suggestion
          %{first_trip | run_id: first_full_trip.run_id}
```